### DISCUSSION TITLE
Debugging Functions Executed in AppWrite Functions

### DISCUSSION THREAD

**Discussion by michaldev** [appwrite/appwrite#5463] [Q&A]
I use AppWrite Cloud. Is debugging functions in Appwrite really as difficult as my initial struggles with the platform?

This is my code:
```
import json

from appwrite.client import Client
from appwrite.services.databases import Databases
from appwrite.permission import Permission
from appwrite.role import Role


def main(req, res):
    print("first")
    data = json.loads(req.payload)
    user_id = req.variables.get('APPWRITE_FUNCTION_USER_ID', None)
    print("second")

    client = Client()
    database = Databases(client)

    (client.set_endpoint(req.variables.get('APPWRITE_FUNCTION_ENDPOINT', None))
     .set_project(req.variables.get('APPWRITE_FUNCTION_PROJECT_ID', None))
     .set_key(req.variables.get('APPWRITE_FUNCTION_API_KEY', None))
     .set_self_signed(True))

    print("third")

    database.create_document(
        'main',
        'accounts',
        user_id,
        data,
        [
            Permission.read(Role.user(user_id)),
            Permission.write(Role.user(user_id)),
            Permission.update(Role.user(user_id)),
        ],
    )
    res.json({"success": True})
```

I don't have any logs or errors at all. Only "Faild" label...
<img width="1158" alt="Zrzut ekranu 2023-04-29 o 17 22 53" src="https://user-images.githubusercontent.com/5132385/235310438-48daa47e-8f10-4a78-9192-aa4f0a1e5ca9.png">

<img width="1164" alt="Zrzut ekranu 2023-04-29 o 17 23 57" src="https://user-images.githubusercontent.com/5132385/235310500-4265ecf6-64e4-4325-85e2-d2b2f90e4bdb.png">

Manually removing line by line, I noticed that the problem lies in this piece of code:
```
    database.create_document(
        'main',
        'accounts',
        user_id,
        data,
        [
            Permission.read(Role.user(user_id)),
            Permission.write(Role.user(user_id)),
            Permission.update(Role.user(user_id)),
        ],
    )
```

If I remove this piece, I get my prints back. 
<img width="1155" alt="Zrzut ekranu 2023-04-29 o 17 28 15" src="https://user-images.githubusercontent.com/5132385/235310681-41975dfd-1eca-4015-aecd-91bc94af0272.png">

My questions:
1. Why am I not seeing any errors?
2. Why am I not seeing prints that are located before the line where the error occurs until I remove the cause of the error?
3. Is this just a Python-specific issue or does it apply to other programming languages as well?

**Comment by squallsama** [comment]
Same applies to dart and kotkin runtimes, so I assume it's for all execution runtimes

**Comment by michaldev** [comment]
Is there any proven method for writing and debugging these functions? So far, I don't know what error I'm getting because I can't see it... The lack of error messages is something new in programming.

**Reply by squallsama** [reply]
> Is there any proven method for writing and debugging these functions? So far, I don't know what error I'm getting because I can't see it... The lack of error messages is something new in programming.

Well...You can remove everything and add line by line into function and deploy it and see what causes the error...This is terrible and I really hope that appwrite will fix error messages and provide a way to how debug functions remotely or locally
